tests/components/dyson/test_fan.py

```diff
@@ -320,14 +384,542 @@ def test_service_set_night_mode(self):
         self.hass.data[DYSON_DEVICES] = []
         dyson_device.entity_id = 'fan.living_room'
         self.hass.data[dyson.DYSON_FAN_DEVICES] = [dyson_device]
-        dyson.setup_platform(self.hass, None, mock.MagicMock())
+        dyson.setup_platform(self.hass, None,
+                             mock.MagicMock(), mock.MagicMock())
 
         self.hass.services.call(dyson.DOMAIN, dyson.SERVICE_SET_NIGHT_MODE,
                                 {"entity_id": "fan.bed_room",
                                  "night_mode": True}, True)
-        assert not dyson_device.night_mode.called
+        assert not dyson_device.set_night_mode.called
 
         self.hass.services.call(dyson.DOMAIN, dyson.SERVICE_SET_NIGHT_MODE,
                                 {"entity_id": "fan.living_room",
                                  "night_mode": True}, True)
-        dyson_device.night_mode.assert_called_with(True)
+        dyson_device.set_night_mode.assert_called_with(True)
+
+
+@mock.patch('libpurecool.dyson.DysonAccount.login', return_value=True)
+@mock.patch('libpurecool.dyson.DysonAccount.devices',
+            return_value=[_get_dyson_purecool_device()])
+async def test_purecool_service_set_night_mode(hass, mocked_devices):
+    """Test set night mode service."""
+    dyson_parent.setup(hass, _get_config())
+    hass.block_till_done()
+    fan = hass.data[dyson.DYSON_FAN_DEVICES][0]
```

Comment:
We shouldn't assert anything about the entity directly. Use the hass core to set up the component and platform and then check the entity state via the core state machine or call services via the core while patching the library methods and assert those.

All these tests need to be changed to not be centered on the entity. We should rely on the core as the source for truth, not the entity. The entity is just an implementation detail.

Reply:
Done. I rewrote all the tests using your advice. I must say that  this way the tests are a lot more readable and easier to write. The new asyncio stuff and the pytest fixtures are great.